pkg/nfd-master/nfd-master.go

```diff
@@ -440,10 +452,120 @@ func (m *nfdMaster) SetLabels(c context.Context, r *pb.SetLabelsRequest) (*pb.Se
+func (m *nfdMaster) setTaints(c context.Context, cli *kubernetes.Clientset, taints []corev1.Taint, nodeName string) error {
+	// Fetch the node object.
+	node, err := m.apihelper.GetNode(cli, nodeName)
+	if err != nil {
+		return err
+	}
+
+	// De-serialize the taints annotation into Taint type for comparision below.
+	annotations := node.GetAnnotations()
+	oldTaints := []corev1.Taint{}
+	if val, ok := annotations[nfdv1alpha1.NodeTaintsAnnotation]; ok {
+		sts := strings.Split(val, ",")
+		corev1Taints, _, err := taintutils.ParseTaints(sts)
+		if err != nil {
+			return err
+		}
+		for _, ct := range corev1Taints {
+			oldTaints = append(oldTaints,
+				corev1.Taint{
+					Key:    ct.Key,
+					Value:  ct.Value,
+					Effect: ct.Effect})
+		}
+	}
+
+	// Find out the diff between the old (from taints annotation) and
+	// new taints (from NodeFeatureRule) from which we get the taints
+	// to remove from the node.
+	taintsToRemove := []corev1.Taint{}
+	for _, taint := range oldTaints {
+		if !taintutils.TaintExists(taints, &taint) {
+			taintsToRemove = append(taintsToRemove, taint)
+		}
+	}
+
+	// We have taintsToRemove that we found from the annotation and here
+	// we first check if the taints still exist on the node and if yes
+	// remove it, otherwise move on.
+	for _, taintToRemove := range taintsToRemove {
+		if taintutils.TaintExists(node.Spec.Taints, &taintToRemove) {
+			newTaints, removed := taintutils.DeleteTaint(node.Spec.Taints, &taintToRemove)
+			if !removed {
+				klog.V(1).Infof("taint %q already deleted from node", taintToRemove.ToString())
+			}
+			node.Spec.Taints = newTaints
+		}
+	}
+
+	// Add new taints.
+	if len(taints) > 0 {
+		var newNode *corev1.Node
+		oldNode := node
+		oldNodeCopy := oldNode
+		for _, taint := range taints {
+			curNewNode, _, err := taintutils.AddOrUpdateTaint(oldNodeCopy, &taint)
+			if err != nil {
+				return fmt.Errorf("failed to add %q taint on node %v", taint, node.Name)
+			}
+			newNode = curNewNode
+			oldNodeCopy = curNewNode
+			continue
+		}
+		err = controller.PatchNodeTaints(c, cli, nodeName, oldNode, newNode)
+		if err != nil {
+			return fmt.Errorf("failed to patch the node %v", node.Name)
+		}
+
+	}
+
+	// Serialize the new taints into string and update the annotation
+	// with that string.
+	if len(taints) > 0 {
+		taintStrs := make([]string, 0, len(taints))
+		for _, taint := range taints {
+			taintStrs = append(taintStrs, taint.ToString())
+		}
+		annotations[nfdv1alpha1.NodeTaintsAnnotation] = strings.Join(taintStrs, ",")
+	} else {
+		delete(annotations, nfdv1alpha1.NodeTaintsAnnotation)
+	}
+	node.SetAnnotations(annotations)
+
+	// Update the node only if there is stale tains to remove or new tains to add.
+	if len(taintsToRemove) > 0 || len(taints) > 0 {
+		klog.Infof("received tainting request for node %q", nodeName)
+		err = m.apihelper.UpdateNode(cli, node)
+		if err != nil {
+			return fmt.Errorf("failed to taint the node %q: %v", node.Name, err)
+		}
+	}
```

Comment:
Now that `PatchNodeTaints()` handles update of the node taints we only need to update the annotation here. We can simply compare the old and new annotation and patch node if a change is needed. We should not update e.g. taints anymore in this phase

In pseudo-code something like this

```
if oldAnnotation != newAnnotation {
    createJsonPatchForAnnotation()
    patchNode()
}

```

Reply:
Ah right 🤦‍♂️